File: backend/app/schemas/envelope.py
from typing import Generic, TypeVar

from pydantic import BaseModel, Field

DataT = TypeVar("DataT")


class ApiError(BaseModel):
    """统一错误结构，便于前端按 code 和 message 展示失败原因。"""

    code: str
    message: str
    details: dict[str, object] = Field(default_factory=dict)


class ApiEnvelope(BaseModel, Generic[DataT]):
    """统一响应外壳，成功时写 data，失败时写 error。"""

    data: DataT | None = None
    error: ApiError | None = None

File: packages/contracts/src/api/envelope.ts
import { z } from "zod";

export const apiErrorSchema = z.object({
  code: z.string(),
  message: z.string(),
  // 可选 i18n 翻译信息：词条（zh/en 双语）在前端资源，按 key + 参数翻译；
  // message 仅作词条缺失时的兜底，不直接作为界面文案
  i18n: z
    .object({
      key: z.string(),
      params: z.record(z.string(), z.union([z.string(), z.number()])).optional(),
    })
    .optional(),
  details: z.record(z.string(), z.unknown()).default({}),
});

export type ApiError = z.infer<typeof apiErrorSchema>;

export type ApiEnvelope<T> = {
  data: T | null;
  error?: ApiError | null;
};
